src/main/java/software/amazon/awssdk/crt/eventstream/ClientConnection.java

```diff
@@ -0,0 +1,197 @@
+        }
+
+        byte[] headersBuf = headers != null ? Header.marshallHeadersForJNI(headers) : null;
+
+        int result = sendProtocolMessage(getNativeHandle(), headersBuf, payload, messsageType.getEnumValue(), messageFlags, callback);
+
+        if (result != 0) {
+            int errorCode = CRT.awsLastError();
+            throw new CrtRuntimeException(errorCode, CRT.awsErrorString(errorCode));
+        }
+    }
+
+    /**
+     * Create a new stream. Activate() must be called on the stream for it to actually initiate the new stream.
+     * @param continuationHandler handler to process continuation messages and state changes.
+     * @return The new continuation object.
+     */
+    public ClientConnectionContinuation newStream(final ClientConnectionContinuationHandler continuationHandler) {
+        if (isNull()) {
+            throw new IllegalStateException("close() has already been called on this object.");
+        }
+
+        long continuationHandle = newClientStream(getNativeHandle(), continuationHandler);
+
+        if (continuationHandle == 0) {
+            int lastError = CRT.awsLastError();
+            throw new CrtRuntimeException(lastError, CRT.awsErrorString(lastError));
+        }
+
+        ClientConnectionContinuation connectionContinuation = new ClientConnectionContinuation(continuationHandle);
+        continuationHandler.continuation = connectionContinuation;
+
+        return connectionContinuation;
+    }
+
+    /**
+     * Initiates a new outgoing event-stream-rpc connection. The future will be completed once the connection either
+     * succeeds or fails.
+     * @param hostName hostname to connect to, this can be an IPv4 address, IPv6 address, a local socket address, or a
+     *                 dns name.
+     * @param port port to connect to hostName with. For local socket address, this value is ignored.
+     * @param socketOptions socketOptions to use.
+     * @param tlsContext (optional) tls context to use for using SSL/TLS in the connection.
+     * @param bootstrap clientBootstrap object to run the connection on.
+     * @param connectionHandler handler to process connection messages and state changes.
+     * @return The future will be completed once the connection either succeeds or fails.
+     */
+    public static CompletableFuture<Void> connect(final String hostName, short port, final SocketOptions socketOptions,
+                                           final ClientTlsContext tlsContext, final ClientBootstrap bootstrap,
+                                           final ClientConnectionHandler connectionHandler) {
+        long tlsContextHandle = tlsContext != null ? tlsContext.getNativeHandle() : 0;
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        ClientConnectionHandler handlerShim = new ClientConnectionHandler() {
+            @Override
+            protected void onConnectionSetup(ClientConnection connection, int errorCode) {
```

Comment:
I'm debating whether or not it would be better if we did this:

If the connection fails, don't invoke ANY of the handler callbacks, just have the connect() function's future complete exceptionally. Now we can remove the error_code argument from ClientConnectionHandler.onConnectionSetup(). Now none of the handler callbacks fire if the connection failed. Now there's only one way to check for errors instead of 2.

Thoughts?

Reply:
we cannot depend purely on futures, or older android customers will not be able to use CRT.